kube-controllers/pkg/controllers/loadbalancer/loadbalancer_controller.go

```diff
@@ -30,157 +32,143 @@ import (
-
-		// Get all profile objects from Calico datastore.
-		serviceList, err := clientset.CoreV1().Services("").List(ctx, metav1.ListOptions{})
-		if err != nil {
-			return nil, err
-		}
-
-		// Filter out only services of type LoadBalancer
-		for _, svc := range serviceList.Items {
-			if svc.Spec.Type == v1.ServiceTypeLoadBalancer {
-				key := createHandle(&svc)
-				filteredServices[key] = svc
-			}
-		}
-		log.Debugf("Found %d Service LoadBalancer in Calico datastore", len(filteredServices))
-		return filteredServices, nil
-	}
-
-	cacheArgs := rcache.ResourceCacheArgs{
-		ListFunc:    listFunc,
-		ObjectType:  reflect.TypeOf(&v1.Service{}),
-		LogTypeDesc: "Service",
-	}
-	ccache := rcache.NewResourceCache(cacheArgs)
-
-	svcHandler := cache.ResourceEventHandlerFuncs{
-		AddFunc: func(obj interface{}) {
-			log.Debugf("Got ADD event for service")
-			svc := obj.(*v1.Service)
-			if isCalicoManagedLoadBalancer(svc, cfg.AssignIPs) {
-				ccache.Set(createHandle(svc), svc)
+func NewLoadBalancerController(clientset *kubernetes.Clientset, calicoClient client.Interface, cfg config.LoadBalancerControllerConfig, serviceInformer cache.SharedIndexInformer, dataFeed *node.DataFeed) controller.Controller {
+	lbc := &loadBalancerController{
+		calicoClient:     calicoClient,
+		cfg:              cfg,
+		clientSet:        clientset,
+		dataFeed:         dataFeed,
+		syncerUpdates:    make(chan interface{}),
+		syncChan:         make(chan interface{}, 1),
+		ipamBlocks:       make(map[string]model.KVPair),
+		ipPools:          make(map[string]api.IPPool),
+		serviceInformer:  serviceInformer,
+		serviceLister:    v1lister.NewServiceLister(serviceInformer.GetIndexer()),
+		servicesToUpdate: make(map[string]serviceObject),
+	}
+
+	lbc.RegisterWith(lbc.dataFeed)
+	lbc.dataFeed.Start()
+
+	_, err := lbc.serviceInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		DeleteFunc: func(obj interface{}) {
+			if svc, ok := obj.(*v1.Service); ok {
+				if IsCalicoManagedLoadBalancer(svc, cfg.AssignIPs) {
+					lbc.syncerUpdates <- serviceObject{
+						updateType: serviceUpdateTypeDELETE,
+					}
+				}
 			}
 		},
 		UpdateFunc: func(oldObj, newObj interface{}) {
-			log.Debugf("Got UPDATE event for service")
-			log.Debugf("Old object: %v", oldObj)
-			log.Debugf("New object: %v", newObj)
 			svcNew := newObj.(*v1.Service)
 			svcOld := oldObj.(*v1.Service)
-
-			// Service changed type, we need to release the addresses used by the service
+			handle, err := createHandle(svcNew)
+			if err != nil {
+				log.WithError(err).Error("Error creating load balancer handle")
+				return
+			}
 			if svcNew.Spec.Type != v1.ServiceTypeLoadBalancer &&
 				svcOld.Spec.Type == v1.ServiceTypeLoadBalancer {
-				ccache.Delete(createHandle(svcOld))
+				lbc.syncerUpdates <- serviceObject{
+					updateType: serviceUpdateTypeDELETE,
+				}
 			} else if svcOld.Annotations[annotationIpv4Pools] != svcNew.Annotations[annotationIpv4Pools] ||
```

Comment:
I think it's going to be a lot cleaner to keep all of the logic about whether or not we should act on updates within the reconcile loop - when we get an update, let's just send it through as an update. The other side of the channel should be doing all of the processing.

So basically, this branch should look something like this:

```go
UpdateFunc: func(oldObj, newObj interface{}) {
  // Tell the main loop that this service has changed.
  lbc.serviceUpdates <- NamespacedName{Name: newObj.Name, Namespace: newObj.Namespace}
}
```

The main loop can decide if it cares about this update, it can decide to batch actions, etc., etc.

It doesn't even need to cache the object - the informer already has a backing cache that is accessible and updated at this point, so we just need to tell the main loop "something changed, here's the context"

Reply:
@caseydavenport I think that we should do a check here to see if we care about the update. It's mostly because we want to update any IPs if the annotations have changed. If we just trigger an update here without checking if the annotations are different it will be harder for us to know if we should re-assign the IP or the change was not relevant to us. I agree that we do not need to pass the service, but rather just use the informer

Reply:
It might work - I need to think about this one a little bit more. My fear is that comparing the annotations between the "new" and the "old" version of the Service isn't strictly speaking the thing this controller cares about.

The controller wants to compare the annotations, the `LoadBalancer.Ingress` values, and the actual IPAM allocation database to determine whether we need to allocate or release IP addresses.

Comparing the annotations is a proxy for that, and probably catches _most_ cases but isn't guaranteed to catch all of the edge cases we care about.

For example, what happens if another controller modifies the Service and overwrites the status.LoadBalancer values that were previously written by this controller? We could try to write special logic to catch that here, but we may not be able to foresee every scenario that requires an update on the channel.